packages/permission-controller/src/PermissionController.ts

```diff
@@ -1991,6 +2110,214 @@ export class PermissionController<
+   * @param incrementalRequestedPermissions - The requested permissions to merge.
+   * @returns The merged permissions and the resulting diff.
+   */
+  #mergeIncrementalPermissions(
+    existingPermissions: Exclude<
+      ReturnType<typeof this.getPermissions>,
+      undefined
+    >,
+    incrementalRequestedPermissions: RequestedPermissions,
+  ):
+    | [
+        SubjectPermissions<
+          ValidPermission<string, ExtractCaveats<ControllerCaveatSpecification>>
+        >,
+        PermissionDiffMap<string, CaveatConstraint>,
+      ]
+    | [] {
+    const permissionDiffMap: PermissionDiffMap<string, CaveatConstraint> = {};
+
+    // Use immer's produce as a convenience for calculating the new permissions
+    // without mutating the existing permissions or committing the results to state.
+    const newPermissions = immerProduce(
+      existingPermissions,
+      (draftExistingPermissions) => {
+        const leftPermissions =
+          draftExistingPermissions as RequestedPermissions;
+
+        Object.entries(incrementalRequestedPermissions).forEach(
+          ([targetName, rightPermission]) => {
+            if (
+              !hasProperty(
+                // We type-guard the targetName key out of existence without this cast.
+                leftPermissions as Record<string, unknown>,
+                targetName,
+              )
+            ) {
+              const [newPermission, caveatsDiff] = this.#mergePermission(
+                {},
+                rightPermission,
+              );
+
+              leftPermissions[targetName] = newPermission;
+              permissionDiffMap[targetName] = caveatsDiff;
+            } else {
+              const [newPermission, caveatsDiff] = this.#mergePermission(
+                leftPermissions[targetName],
+                rightPermission,
+              );
+
+              if (Object.keys(caveatsDiff).length > 0) {
+                leftPermissions[targetName] = newPermission;
+                permissionDiffMap[targetName] = caveatsDiff;
+              }
+              // Otherwise, leave the left permission as-is; its authority has
+              // not changed.
+            }
+          },
+        );
+      },
+    );
+
+    if (Object.keys(permissionDiffMap).length === 0) {
+      return [];
+    }
+    return [newPermissions, permissionDiffMap];
+  }
+
+  /**
+   * Performs a right-biased union between two permissions. The task of merging caveats
+   * of the same type between the two permissions is delegated to the corresponding
+   * caveat type's merger implementation.
+   *
+   * Throws if the left-hand and right-hand permissions both have a caveat whose
+   * specification does not provide a caveat merger function.
+   *
+   * @param leftPermission - The left-hand permission to merge.
+   * @param rightPermission - The right-hand permission to merge.
+   * @returns The merged permission.
+   */
+  #mergePermission(
```

Comment:
Wondering if it makes sense for this function to have a fast path when `leftPermission` is empty, since that may be a likely scenario.

Reply:
I don't think the performance benefits (if any) would be worth the added complexity of the optimization.